[CI/Build] mypy: Resolve some errors from checking vllm/engine

russellb opened the pull request:
This PR is a chunk of mypy checking progress.

1. I got to where I knew everything would pass when running `mypy` with
   `--follow-imports skip`.  (though some new issues have crept back in since I passed this point)

2. I then started working on expanding the coverage for what can be checked
   using `--follow-imports silent`. Ongoing progress will come in chunks to make it
   easier to review. I started with 100 errors and it's down to 60 right now.

Part of issue  #3680

russellb committed:
[CI/Build]: mypy: drop duplicate coverage

I had all directories listed in `tools/mypy.sh` to ensure they worked
with `--follow-imports skip`. This change includes the last tweaks
necessary for all of those to pass.

As a next step, remove the ones that are already running
with `--follow-imports silent` based on the config in
`pyproject.toml`. Following commits will start resolving errors that
occur when adding more coverage here.

Signed-off-by: Russell Bryant <rbryant@redhat.com>

russellb committed:
[CI/Build] Fix some mypy errors from checking vllm/engine

This fixes 27 of the 100 errors that occur when running:

    mypy --follow-imports silent vllm/engine

It is submitted as a checkpoint to make it easier to review.

Signed-off-by: Russell Bryant <rbryant@redhat.com>

DarkLight1337 commented on vllm/config.py:
```diff
@@ -579,6 +579,10 @@ class CacheConfig:
             profiled num_gpu_blocks if specified. Does nothing if None.
     """
 
+    # Will be set after profiling.
+    num_gpu_blocks: Optional[int] = None
+    num_cpu_blocks: Optional[int] = None
```
Doesn't this set the value at the class level rather than the instance level?

russellb replied:
oops ... not sure what I was thinking here.

Thank you for reviewing so closely! That's why I wanted to send it in smaller chunks to make sure it got a close look.

DarkLight1337 commented on vllm/engine/llm_engine.py:
```diff
@@ -1236,8 +1236,9 @@ def _advance_to_next_step(
-                seq_group.update_num_computed_tokens(
-                    seq_group_metadata.token_chunk_size)
+                token_chunk_size = (seq_group_metadata.token_chunk_size if
+                                    seq_group_metadata.token_chunk_size else 0)
```
```suggestion
                                    seq_group_metadata.token_chunk_size is not None else 0)
```

russellb replied:
same result, just a style preference?

DarkLight1337 replied:
It's to emphasize that `seq_group_metadata.token_chunk_size` can be `None` (since it's not clear from just reading on GitHub).

DarkLight1337 reviewed (approved):
Overall looks ok, PTAL at my comments though.

njhill commented on vllm/engine/output_processor/util.py:
```diff
@@ -1,13 +1,11 @@
 
 def create_output_by_sequence_group(
-        outputs: GenericSequence[Union[SamplerOutput, PoolerOutput]],
+        outputs: GenericSequence[Iterable],
```
This doesn't look right to me ... outputs shouldn't be a sequence of iterables?

russellb replied:
`SamplerOutput` and `PoolerOutput` are iterable, I believe. This was just a simpler way to express the requirements here since the code doesn't actually care about any details of the type otherwise

DarkLight1337 replied:
Actually, my IDE (VSCode) flags `SamplerOutput` and `PoolerOutput` as not being iterable. We can solve this by adding a type annotation like this:

```
class SamplerOutput(...):

    if TYPE_CHECKING:
        def __iter__(self) -> Iterator[CompletionSequenceGroupOutput]:
            raise RuntimeError("Method should be automatically generated")
```

russellb replied:
interesting - i thought it made mypy happy, apologies! I'll update this.

DarkLight1337 replied:
My IDE uses pyright for linting so it may actually pass on mypy.

russellb replied:
I changed how this works. Can you see if it works for you @DarkLight1337 ?

DarkLight1337 replied:
Yes, this looks good, thanks!

njhill commented on vllm/engine/multiprocessing/client.py:
```diff
@@ -630,7 +635,8 @@ async def _process_request(
         if isinstance(params, SamplingParams) and \
-            params.guided_decoding is not None:
+            params.guided_decoding is not None and \
+            self.decoding_config and lora_request:
```
`lora_request` can be None here.

Also instead of skipping we should probably pass default_guided_backend as
```python
default_guided_backend = self.decoding_config.guided_decoding_backend if self.decoding_config else DecodingConfig.guided_decoding_backend
```

russellb replied:
appllied your suggestions

njhill commented on vllm/core/scheduler.py:
```diff
@@ -4,8 +4,9 @@
-from typing import (Callable, Deque, Dict, Iterable, List, Optional, Set,
-                    Tuple, Union)
+from typing import Callable, Deque, Dict, Iterable, List, Optional
+from typing import Sequence as SequenceType
```
Elsewhere we use `GenericSequence` as the alias for this

russellb replied:
ah, sure enough, thanks.

njhill commented on vllm/engine/output_processor/stop_checker.py:
```diff
@@ -57,7 +57,8 @@ def maybe_stop_sequence(
         # This assumes a single token produced per step.
         last_token_id = seq.get_last_token_id()
-        if last_token_id in sampling_params.stop_token_ids:
+        if sampling_params.stop_token_ids and \
+            last_token_id in sampling_params.stop_token_ids:
```
nit: bit simpler
```suggestion
        if last_token_id in (sampling_params.stop_token_ids or ()):
```

russellb replied:
good suggestion, thanks!

DarkLight1337 replied:
Maybe add a method to `sampling_params` like `has_stop_token`? (Although this seems to only be used here...)

russellb replied:
true, but it would read nicer in this spot! :-)

russellb replied:
I applied the suggestion from @njhill here

njhill reviewed (commented):
Thanks @russellb!

russellb commented:
@njhill @DarkLight1337 Thank you both for the close review! I will let you know once I have addressed the feedback.

russellb committed:
mypy: avoid using `backend` as two different types

`backend` was used in two different places in this function as two
different types. There was no error here, but `mypy` didn't like it.

Signed-off-by: Russell Bryant <rbryant@redhat.com>

russellb committed:
mypy: Add type validation to support_torch_compile()

The code here previously raised an error in mypy because the `forward`
method may not exist on the decorated class.  Add explicit validation
of this requirement.

Signed-off-by: Russell Bryant <rbryant@redhat.com>

russellb committed:
mypy: Ignore type warnings when overriding methods in decorator

In the support_torch_compile() decorator, mypy doesn't like the lines
overriding class methods. The code looks fine to me, so just ignore
the warnings here.

Signed-off-by: Russell Bryant <rbryant@redhat.com>

russellb committed:
mypy: Assert some type assumptions in output processing

Several places there are assumptions that a type is a more specific
type than the generic one specified in function parameters. The
parameters are set because it's based on a more generic parent class,
so we can't override them here.

Signed-off-by: Russell Bryant <rbryant@redhat.com>

russellb committed:
mypy: ensure we don't try to add None to a number

mypy pointed out that a couple of variables are Optional (could be
None), but we are adding them to a float or int. Ensure that we don't
try to add None.

Signed-off-by: Russell Bryant <rbryant@redhat.com>

russellb commented:
marked as a draft to fix some things - will mark ready for review again when done

russellb commented:
looks like there's a real failure in a metrics test. I'm looking at it now.

DarkLight1337 reviewed (approved):
Should be good to go now, thanks again for your efforts!

njhill commented on vllm/outputs.py:
```diff
@@ -92,7 +93,7 @@ class RequestOutput:
         self,
         request_id: str,
-        prompt: Optional[str],
+        prompt: Optional[PromptType],
```
Sorry for the after-the-fact comment and for missing in review .. I think this one might not be the right change ... it breaks things in other places and is also part of the API so probably shouldn't change in this way.

I think probably the reason it was changed was due to the recent beam_search logic which still needs a lot of work and I think incorrectly is assigning a var of PromptType to this.

I can address this in another PR that I'm working on right now.

russellb replied:
@njhill sorry about that. I think it was OK at one point, and then when I rebased on `main` some new errors started happening and I didn't fix them before merge. Let me know if you'd like me to just revert this if you want it fixed quickly.

njhill replied:
Thanks @russellb no problem I have a PR almost ready
